### ISSUE TITLE
Error when loading test.html using MapStaticAssets: Failed to load resource: net::ERR_CONTENT_DECODING_FAILED

### ISSUE THREAD

**Issue by danroth27** [dotnet/aspnetcore#58940]
Repro steps:

- Create an empty ASP.NET Core app
- Add `app.MapStaticAssets()`
- Add wwwroot/test.html

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <title></title>
</head>
<body>
    <h1>Test</h1>
</body>
</html>
```

Run the app and browser to /test.html

Expected result: test.html renders in the browser

Actual result: Failed to load resource: net::ERR_CONTENT_DECODING_FAILED

Workaround: Swap `app.MapStaticAssets()` for `app.UseStaticFiles()`

**Comment by space-alien** [comment]
This is a pain.

I have found that the presence of a `<body>` tag within the html file somehow breaks the serving of the file.

OK:
```html
<!DOCTYPE HTML>

<html>

This file is seemingly served correctly.

</html>
```

Broken:
```html
<!DOCTYPE HTML>

<html>

<body>
    This file won't render in the browser. Dev tools report "Failed to load
    response data: No data found for resource with given identifier".
    Yum.
</body>

</html>
```

Hope this helps.

**Comment by fingers10** [comment]
> This is a pain.
> 
> I have found that the presence of a `<body>` tag within the html file somehow breaks the serving of the file.
> 
> OK:
> 
> <!DOCTYPE HTML>
> 
> <html>
> 
> This file is seemingly served correctly.
> 
> </html>
> Broken:
> 
> <!DOCTYPE HTML>
> 
> <html>
> 
> <body>
>     This file won't render in the browser. Dev tools report "Failed to load
>     response data: No data found for resource with given identifier".
>     Yum.
> </body>
> 
> </html>
> Hope this helps.

by default blazor wasm apps will serve `index.html` file and is that why my blazor wasm apps are not getting served? @danroth27 any helps?

**Comment by dashiell-zhang** [comment]
w

> Repro steps:
> 
> * Create an empty ASP.NET Core app
> * Add `app.MapStaticAssets()`
> * Add wwwroot/test.html
> 
> <!DOCTYPE html>
> <html>
> <head>
>     <meta charset="utf-8" />
>     <title></title>
> </head>
> <body>
>     <h1>Test</h1>
> </body>
> </html>
> Run the app and browser to /test.html
> 
> Expected result: test.html renders in the browser
> 
> Actual result: Failed to load resource: net::ERR_CONTENT_DECODING_FAILED
> 
> Workarounds:
> 
> * Swap `app.MapStaticAssets()` for `app.UseStaticFiles()`
> * Use a .json file instead of a .html file

I also encountered the same problem. I temporarily solved this problem by adding the following information to the project.
```
<ItemGroup>
<Content Remove="wwwroot\test.html" />
<None Include="wwwroot\test.html" />
</ItemGroup>
```
It seems that visual studio performed a gzip encoding on the test.html file during the generation process.

**Comment by danroth27** [comment]
@javiercn

**Comment by Pragmateek** [comment]
This bug is still crawling, I have the exact same scenario and symptoms as @danroth27.
With `MapStaticAssets` Kestrel's response add a `Content-Encoding: gzip` header.
With `UseStaticFiles` this header is not specified at all.

**Comment by javiercn** [comment]
Setting `"ReloadStaticAssetsAtRuntime": "false"` should workaround the issue.

**Comment by SteveSandersonMS** [comment]
@javiercn Will we be fixing this in a patch?

**Comment by danroth27** [comment]
This issue appears to only happen in VS with browser link enabled. We think the issue is due to how browser link intercepts the response and tries to inject content into it.

**Comment by dashiell-zhang** [comment]
> This issue appears to only happen in VS with browser link enabled. We think the issue is due to how browser link intercepts the response and tries to inject content into it.

@danroth27 Hello，This problem exists not only when Visual Studio is running. After the code is released and deployed in nginx, net::ERR_CONTENT_DECODING_FAILED will still appear when accessing the corresponding test.html from other computers.

**Comment by danroth27** [comment]
@fingers10 Thanks for these additional details! That's probably because of a similar issue where hot reload is trying to inject a script on to the page.

**Comment by danroth27** [comment]
> [@danroth27](https://github.com/danroth27) Hello, This problem exists not only when Visual Studio is running. After the code is released and deployed in nginx, net::ERR_CONTENT_DECODING_FAILED will still appear when accessing the corresponding test.html from other computers.

@fingers10 Hmm, that is odd. Could this possibly be due to a misconfiguration of the compression settings in nginx?

**Comment by fingers10** [comment]
> [@fingers10](https://github.com/fingers10) Thanks for these additional details! That's probably because of a similar issue where hot reload is trying to inject a script on to the page.

@danroth27 Any suggestion / workaround for this? its difficult without hot reload

**Comment by danroth27** [comment]
> [@danroth27](https://github.com/danroth27) Any suggestion / workaround for this? its difficult without hot reload

See https://github.com/dotnet/aspnetcore/issues/58940#issuecomment-2631386964

**Comment by fingers10** [comment]
> Setting `"ReloadStaticAssetsAtRuntime": "false"` should workaround the issue.

@danroth27 @javiercn I'm facing this issue with standalone blazor wasm app. Where should I add this setting in blazor wasm standalone app?

**Comment by Mohamed1984** [comment]
> > Setting `"ReloadStaticAssetsAtRuntime": "false"` should workaround the issue.
> 
> [@danroth27](https://github.com/danroth27) [@javiercn](https://github.com/javiercn) I'm facing this issue with standalone blazor wasm app. Where should I add this setting in blazor wasm standalone app?

In lauchSettings.json

Anyway, it seems the issue is due to gzip compression of static assets collides with browser link. You can stop hot reload by adding the following line in launchSettings.json
"hotReloadEnabled": true


`"https": {
    "commandName": "Project",
    "dotnetRunMessages": true,
    "launchBrowser": true,
    "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
    "applicationUrl": "https://localhost:5032",
    "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
    },
    "hotReloadEnabled": true
},`

**Comment by fingers10** [comment]
@Mohamed1984 Many thanks for the support and help.

@danroth27 @javiercn , I'm surprised.. 

1. I added `"reloadStaticAssetsAtRuntime": "false"` in `launchSettings.json` and did `dotnet watch run`
2. Now I stopped the app and changed the value to `"reloadStaticAssetsAtRuntime": "true"` and did `dotnet watch run` 
3. Now I stopped the app and then removed `"reloadStaticAssetsAtRuntime": "false"` from `launchSettings.json` and  did `dotnet watch run` 

Now hot reload is working without any issues even if I don't have `"reloadStaticAssetsAtRuntime": "false"` in `launchSettings.json`

**Comment by andrexx** [comment]
ASP.NET MVC app hosted in IIS. Put googleXXXX.html file to wwwroot folder to site verification. The same issue - can't open the file until rebuilding and redeploying the app with file.
I think this middleware ignores files not included in *.staticwebassets.endpoints.json

Another issue is that IIS doesn't serve compressed files, there is a lack of documentation, and compression is useless by default even if you use UseStaticFiles.

I'm going to remove all usages of MapStaticAssets and switch back to UseStaticFiles. It looks like this feature is not ready for production use.

**Comment by pmarangoni** [comment]
I switched from using UseStaticFiles to MapStaticAssets a few weeks ago. My site (served from IIS on Windows Server) is now returning 404 for an html file in the wwwroot folder. This file needs to be accessible for a verification token with Twilio. Is there any way to have it work with MapStaticAssets or not?

**Comment by MackinnonBuck** [comment]
Fixed in https://github.com/dotnet/sdk/pull/50535
